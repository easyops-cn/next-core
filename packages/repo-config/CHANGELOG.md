# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.0.2](https://github.com/easyops-cn/next-core/compare/@next-core/repo-config@1.0.1...@next-core/repo-config@1.0.2) (2023-03-22)

**Note:** Version bump only for package @next-core/repo-config

## [1.0.1](https://github.com/easyops-cn/next-core/compare/@next-core/repo-config@1.0.0...@next-core/repo-config@1.0.1) (2023-03-22)

**Note:** Version bump only for package @next-core/repo-config

# [1.0.0](https://github.com/easyops-cn/next-core/compare/@next-core/repo-config@0.2.5...@next-core/repo-config@1.0.0) (2023-03-22)

### Bug Fixes

- refine i18n ([60966a7](https://github.com/easyops-cn/next-core/commit/60966a747d2e04e72ad6ec7a82251d1b94a50a07))

### Features

- support create-api-ask ([cd18f8e](https://github.com/easyops-cn/next-core/commit/cd18f8e9d3df2676105438d7f772b713b615cf2a))

## [0.2.5](https://github.com/easyops-cn/next-core/compare/@next-core/repo-config@0.2.4...@next-core/repo-config@0.2.5) (2022-12-09)

**Note:** Version bump only for package @next-core/repo-config

## [0.2.4](https://github.com/easyops-cn/next-core/compare/@next-core/repo-config@0.2.3...@next-core/repo-config@0.2.4) (2022-11-07)

**Note:** Version bump only for package @next-core/repo-config

## [0.2.3](https://github.com/easyops-cn/next-core/compare/@next-core/repo-config@0.2.2...@next-core/repo-config@0.2.3) (2022-11-04)

**Note:** Version bump only for package @next-core/repo-config

## [0.2.2](https://github.com/easyops-cn/next-core/compare/@next-core/repo-config@0.2.1...@next-core/repo-config@0.2.2) (2022-11-04)

**Note:** Version bump only for package @next-core/repo-config

## [0.2.1](https://github.com/easyops-cn/next-core/compare/@next-core/repo-config@0.2.0...@next-core/repo-config@0.2.1) (2021-02-07)

### Bug Fixes

- using GPL license ([f9e7bd9](https://github.com/easyops-cn/next-core/commit/f9e7bd9))

# [0.2.0](https://github.com/easyops-cn/next-core/compare/@next-core/repo-config@0.1.2...@next-core/repo-config@0.2.0) (2021-02-05)

### Features

- support to ignore post-build micro-apps ([c1b5e55](https://github.com/easyops-cn/next-core/commit/c1b5e5510c8184cdf0af595eafe9827ba25c43b1))
- support to serve standalone ([c4e20a8](https://github.com/easyops-cn/next-core/commit/c4e20a830e56b5bcbdbc38eb682aee5405471c7d))

## [0.1.2](https://github.com/easyops-cn/next-core/compare/@next-core/repo-config@0.1.1...@next-core/repo-config@0.1.2) (2021-01-26)

**Note:** Version bump only for package @next-core/repo-config

## 0.1.1 (2021-01-22)

### Features

- rename [@easyops](https://github.com/easyops) by [@next-core](https://github.com/next-core), and [@dll](https://github.com/dll) by [@next-dll](https://github.com/next-dll) ([d87ed1e](https://github.com/easyops-cn/next-core/commit/d87ed1e))
