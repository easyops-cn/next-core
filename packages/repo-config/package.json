{
  "name": "@next-core/repo-config",
  "version": "0.2.5",
  "description": "> TODO: description",
  "homepage": "https://github.com/easyops-cn/next-core/tree/master/packages/repo-config",
  "license": "GPL-3.0",
  "main": "index.js",
  "typings": "index.d.ts",
  "repository": {
    "type": "git",
    "url": "git@github.com:easyops-cn/next-core.git"
  },
  "scripts": {
    "test": "cross-env NODE_ENV='test' jest --passWithNoTests",
    "test:ci": "cross-env NODE_ENV='test' CI=true jest --passWithNoTests"
  }
}
