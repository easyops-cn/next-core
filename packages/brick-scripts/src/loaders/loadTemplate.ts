import fs from "fs-extra";
import path from "path";
import { exec } from "child_process";
import klawSync from "klaw-sync";
import * as changeCase from "change-case";
import rp from "request-promise-native";
import { getEasyopsConfig } from "@next-core/repo-config";
import { FileWithContent, TargetType } from "../interface";

// `tsc` will compile files which `import` or `require`,
// thus, we read file content instead of importing.
const packageJson = JSON.parse(
  fs.readFileSync(path.join(__dirname, "../../package.json"), "utf8")
);
const { templateDependencies } = packageJson;
const brickContainerVersion =
  templateDependencies["@next-core/brick-container"];

const workspacePackageJson = JSON.parse(
  fs.readFileSync(path.join(process.cwd(), "package.json"), "utf8")
);
const workspaceHomepage = workspacePackageJson.homepage;
const workspaceLicense = workspacePackageJson.license || "UNLICENSED";
let userName = "bot";
exec("git config user.name", { encoding: "utf8" }, function (err, value) {
  if (err) return;
  userName = value.trim();
});

function escapeRegExp(str: string): string {
  return str.replace(/([.*+?^=!:${}()|[\]/\\])/g, "\\$1");
}

function replaceAll(str: string, find: string, replace: string): string {
  return str.replace(new RegExp(escapeRegExp(find), "g"), replace);
}

function replaceFileContent(
  filePath: string,
  translations: Record<string, string>
): string {
  const content = fs.readFileSync(filePath, "utf8");
  const contentFinds = Object.keys(translations).filter((key) =>
    content.includes(key)
  );
  if (contentFinds.length > 0) {
    return contentFinds.reduce(
      (acc, find) => replaceAll(acc, find, translations[find]),
      content
    );
  }
  return content;
}

export async function loadTemplate({
  targetType,
  packageName,
  brickName,
  processorName,
  targetRoot,
}: {
  targetType: TargetType;
  packageName: string;
  brickName: string;
  processorName: string;
  targetRoot: string;
}): Promise<FileWithContent[]> {
  const { useLocalSdk, usePublicScope } = getEasyopsConfig();

  const npmScopeOfBricks = usePublicScope ? "@next-bricks" : "@bricks";
  const npmScopeOfMicroApps = usePublicScope
    ? "@next-micro-apps"
    : "@micro-apps";
  const npmScopeOfTemplates = usePublicScope
    ? "@next-legacy-templates"
    : "@templates";
  const npmScopeOfLibs = usePublicScope ? "@next-libs" : "@libs";
  const npmScopeOfSdk = usePublicScope ? "@next-sdk" : "@sdk";
  const repoOrgUrl = usePublicScope
    ? "https://github.com/easyops-cn"
    : "https://git.easyops.local/anyclouds";
  const repoGitUrl = usePublicScope
    ? "github.com:easyops-cn"
    : "git.easyops.local:anyclouds";
  const npmRegistryUrl = usePublicScope
    ? "https://registry.npmjs.org"
    : "https://registry.npm.easyops.local";

  const targetMap: { [key: string]: string } = {
    [TargetType.A_NEW_BRICK]: "brick",
    [TargetType.A_NEW_EDITOR_BRICK]: "editor-brick",
    [TargetType.A_NEW_CUSTOM_TEMPLATE]: "custom-template",
    [TargetType.A_NEW_PACKAGE_OF_BRICKS]: "bricks-pkg",
    [TargetType.A_NEW_PACKAGE_OF_LIBS]: "libs-pkg",
    [TargetType.A_NEW_CUSTOM_PROVIDER]: "custom-provider",
    [TargetType.A_NEW_CUSTOM_PROCESSOR]: "custom-processor",
    [TargetType.A_NEW_PACKAGE_OF_PROVIDERS]: "providers-pkg",
    [TargetType.A_NEW_PACKAGE_OF_DLL]: "dll-pkg",
  };
  const templateDirOfFileName = targetMap[targetType];
  const templateRoot = path.join(__dirname, "../../template");
  let templateDir = path.join(templateRoot, templateDirOfFileName);

  const ignores = [".DS_Store"];
  let sdkName: string;
  if (targetType === TargetType.A_NEW_PACKAGE_OF_PROVIDERS) {
    // Providers 库直接使用构件库的模板文件，但忽略其中部分文件。
    templateDir = path.join(
      templateRoot,
      targetMap[TargetType.A_NEW_PACKAGE_OF_BRICKS]
    );
    // Providers 库不需要 `src` 及 `webpack.editors.js`。
    ignores.push("template/bricks-pkg/src/");
    ignores.push("template/bricks-pkg/webpack.editors.js");
    sdkName = `${packageName.replace(/^providers-of-/, "")}-sdk`;
  }

  const translations: Record<string, string> = {
    "$npm-scope-of-bricks$": npmScopeOfBricks,
    "$npm-scope-of-micro-apps$": npmScopeOfMicroApps,
    "$npm-scope-of-templates$": npmScopeOfTemplates,
    "$npm-scope-of-libs$": npmScopeOfLibs,
    "$npm-scope-of-sdk$": npmScopeOfSdk,
    "$repo-org-url$": repoOrgUrl,
    "$repo-git-url$": repoGitUrl,
    "$workspace-homepage$": workspaceHomepage,
    "$kebab-package-name$": packageName,
    $PascalPackageName$: changeCase.pascalCase(packageName),
    "$Title Package Name$": changeCase.capitalCase(packageName),
    $CONSTANT_PACKAGE_NAME$: changeCase.constantCase(packageName),
    $camelCasePackageName$: changeCase.camelCase(packageName),
    $PascalBrickName$: changeCase.pascalCase(brickName),
    "$kebab-brick-name$": `${packageName}.${brickName}`,
    "$kebab-brick-last-name$": brickName,
    "$kebab-custom-provider-name$": `${packageName}.provider-${brickName}`,
    "$generator.version$": `v${packageJson.version}`,
    "$brick.container.version$": brickContainerVersion,
    "$kebab-sdk-name$": sdkName,
    "$kebab-username$": userName,
    $camelProcessorName$: changeCase.camelCase(processorName),
    "$open-source-license$": workspaceLicense,
    "$brick-package-type$":
      targetType === TargetType.A_NEW_PACKAGE_OF_PROVIDERS
        ? "providers"
        : "bricks",
  };

  const filter = (src: string): boolean =>
    ignores.every((item) => !src.includes(item));

  const templateGroups: any = [
    {
      templateDir,
      targetDir: targetRoot,
      files: klawSync(templateDir, {
        depthLimit: 4,
        nodir: true,
        filter: (item) => filter(item.path),
      }),
    },
  ];

  if (targetType === TargetType.A_NEW_PACKAGE_OF_BRICKS && brickName) {
    // Also create a new brick for the new bricks-package
    const brickTemplateDir = path.join(templateRoot, "brick");
    templateGroups.push({
      templateDir: brickTemplateDir,
      targetDir: path.join(targetRoot, "src", brickName),
      files: klawSync(brickTemplateDir, {
        depthLimit: 2,
        nodir: true,
        filter: (item) => filter(item.path),
      }),
    });
  } else if (targetType === TargetType.A_NEW_PACKAGE_OF_PROVIDERS) {
    // Providers 库还有专属的一些模板文件。
    const brickTemplateDir = path.join(templateRoot, "providers-pkg");
    templateGroups.push({
      templateDir: brickTemplateDir,
      targetDir: targetRoot,
      files: klawSync(brickTemplateDir, {
        depthLimit: 2,
        nodir: true,
        filter: (item) => filter(item.path),
      }),
    });

    let sdkVersion = "1.0.0";
    const getVersion = async () => {
      try {
        const sdkPackage = await rp({
          url: `${npmRegistryUrl}/${npmScopeOfSdk}/${sdkName}`,
          json: true,
          strictSSL: false,
        });
        sdkVersion = `^${sdkPackage["dist-tags"].latest}`;
      } catch {
        sdkVersion = "FETCH LATEST VERSION ERROR";
      }
    };

    if (!useLocalSdk) {
      await getVersion();
    } else {
      const sdkPackagePath = path.resolve(
        targetRoot,
        "..",
        "..",
        "sdk",
        sdkName,
        "package.json"
      );
      if (fs.existsSync(sdkPackagePath)) {
        sdkVersion = JSON.parse(
          fs.readFileSync(sdkPackagePath, "utf8")
        ).version;
      }
    }

    translations["$sdk.version$"] = sdkVersion;
  }

  const files: FileWithContent[] = templateGroups.reduce(
    (acc: FileWithContent[], group: any) =>
      acc.concat(
        group.files.map((file: any) => {
          const content = replaceFileContent(file.path, translations);

          let realPath = file.path;
          const basename = path.basename(file.path);
          const filenameFinds = Object.keys(translations).filter((key) =>
            basename.includes(key)
          );
          if (filenameFinds.length > 0) {
            realPath = path.join(
              path.dirname(realPath),
              filenameFinds.reduce(
                (acc, find) => replaceAll(acc, find, translations[find]),
                basename
              )
            );
          }

          return [
            path.join(
              group.targetDir,
              path.relative(group.templateDir, realPath)
            ),
            content,
          ];
        })
      ),
    []
  );

  if (
    [
      TargetType.A_NEW_PACKAGE_OF_BRICKS,
      TargetType.A_NEW_PACKAGE_OF_LIBS,
      TargetType.A_NEW_PACKAGE_OF_PROVIDERS,
      TargetType.A_NEW_PACKAGE_OF_DLL,
    ].includes(targetType)
  ) {
    files.push([
      path.join(targetRoot, "package.json"),
      replaceFileContent(
        path.join(templateRoot, `${targetMap[targetType]}.json`),
        translations
      ),
    ]);
  }

  return files;
}
