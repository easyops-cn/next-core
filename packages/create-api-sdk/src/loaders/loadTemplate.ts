import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import klawSync from "klaw-sync";
import * as changeCase from "change-case";
import { getEasyopsConfig } from "@next-core/repo-config";
import { FileWithContent } from "../interface.js";
import __dirname from "./__dirname.js";

// `tsc` will compile files which `import` or `require`,
// thus, we read file content instead of importing.
const packageJson = JSON.parse(
  fs.readFileSync(path.join(__dirname, "../../package.json"), "utf8")
);

const workspacePackageJson = JSON.parse(
  fs.readFileSync(
    path.join(
      process.env.NODE_ENV === "test"
        ? path.join(__dirname, "../../../..")
        : process.cwd(),
      "package.json"
    ),
    "utf8"
  )
);

const workspaceHomepage = workspacePackageJson.homepage;
const workspaceLicense = workspacePackageJson.license || "UNLICENSED";

function escapeRegExp(str: string): string {
  return str.replace(/([.*+?^=!:${}()|[\]/\\])/g, "\\$1");
}

function replaceAll(str: string, find: string, replace: string): string {
  return str.replace(new RegExp(escapeRegExp(find), "g"), replace);
}

function replaceFileContent(
  filePath: string,
  translations: Record<string, string>
): string {
  const content = fs.readFileSync(filePath, "utf8");
  const contentFinds = Object.keys(translations).filter((key) =>
    content.includes(key)
  );
  if (contentFinds.length > 0) {
    return contentFinds.reduce(
      (acc, find) => replaceAll(acc, find, translations[find]),
      content
    );
  }
  return content;
}

function replaceDepsVersion(jsonString: string, sdkVersion: string): string {
  const pkg = JSON.parse(jsonString);
  if (sdkVersion) {
    pkg.version = sdkVersion;
  }
  return JSON.stringify(pkg, null, 2) + os.EOL;
}

export function loadTemplate(
  serviceName: string,
  sdkRoot: string,
  sdkVersion: string
): FileWithContent[] {
  const { usePublicScope } = getEasyopsConfig();

  const templateDir = path.join(__dirname, "../../template");

  const ignores = [".DS_Store"];
  const filter = (src: string): boolean =>
    ignores.some((item) => !src.includes(item));

  const templateFiles = klawSync(templateDir, {
    depthLimit: 2,
    nodir: true,
    filter: (item) => filter(item.path),
  });

  const translations: Record<string, string> = {
    "$npm-scope-of-sdk$": usePublicScope ? "@next-api-sdk" : "@api-sdk",
    "$workspace-homepage$": workspaceHomepage,
    "$kebab-service-name$": changeCase.paramCase(serviceName),
    "$Title Service Name$": changeCase.capitalCase(serviceName),
    $PascalServiceName$: changeCase.pascalCase(serviceName),
    "$generator.version$": `v${packageJson.version}`,
    "$open-source-license$": workspaceLicense,
  };

  const files: FileWithContent[] = templateFiles.map((file) => {
    const content = replaceFileContent(file.path, translations);
    return [path.join(sdkRoot, path.relative(templateDir, file.path)), content];
  });

  files.push([
    path.join(sdkRoot, "package.json"),
    replaceDepsVersion(
      replaceFileContent(
        path.join(templateDir, "../template.json"),
        translations
      ),
      sdkVersion
    ),
  ]);

  return files;
}
