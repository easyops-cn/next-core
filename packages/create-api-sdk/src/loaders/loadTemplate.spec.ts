import path from "node:path";
import { describe, it, expect } from "@jest/globals";
import { getEasyopsConfig } from "@next-core/repo-config";
import { loadTemplate } from "./loadTemplate.js";

jest.mock("@next-core/repo-config");
jest.mock("./__dirname.js");

describe("loadTemplate", () => {
  it("should work", () => {
    (getEasyopsConfig as jest.Mock).mockReturnValue({
      usePublicScope: false,
    });
    let files = loadTemplate("cd", "/tmp/sdk", "1.2.3");
    files = files.filter(([filePath]) => {
      const fileName = path.basename(filePath);
      return !["README.md"].includes(fileName);
    });
    expect(files).toMatchSnapshot();
  });

  it("should work for public scope", () => {
    (getEasyopsConfig as jest.Mock).mockReturnValue({
      usePublicScope: true,
    });
    let files = loadTemplate("cd", "/tmp/sdk", "1.2.3");
    files = files.filter(([filePath]) => {
      const fileName = path.basename(filePath);
      return !["README.md"].includes(fileName);
    });
    expect(files).toMatchSnapshot();
  });
});
