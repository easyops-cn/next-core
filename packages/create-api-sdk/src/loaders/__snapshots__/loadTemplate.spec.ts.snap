// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadTemplate should work 1`] = `
[
  [
    "/tmp/sdk/src/wrapper.ts",
    "export interface ResponseBodyWrapper<T> {
  code: number;
  data: T;
  error?: string;
  message?: string;
}

export interface ResponseListWrapper<T> {
  list: T[];
  page: number;
  page_size: number;
  total: number;
}
",
  ],
  [
    "/tmp/sdk/tsconfig.json",
    "{
  "extends": "../../tsconfig.module.json",
  "include": ["./src"]
}
",
  ],
  [
    "/tmp/sdk/package.json",
    "{
  "name": "@api-sdk/cd-sdk",
  "version": "1.2.3",
  "description": "API SDK for Cd",
  "homepage": "https://github.com/easyops-cn/next-core/tree/master/sdk/cd-sdk",
  "license": "GPL-3.0",
  "type": "module",
  "main": "dist/cjs/index.js",
  "module": "dist/esm/index.js",
  "typings": "dist/types/index.d.ts",
  "files": [
    "dist",
    "contracts.json"
  ],
  "exports": {
    ".": {
      "types": "./dist/types/index.d.ts",
      "import": "./dist/esm/index.js",
      "require": "./dist/cjs/index.js"
    },
    "./package.json": "./package.json"
  },
  "scripts": {
    "clean": "rimraf dist",
    "prestart": "npm run clean",
    "start": "concurrently -k -n types,main \\"npm run start:types\\" \\"npm run start:main\\"",
    "start:main": "cross-env NODE_ENV=development build-next-libs --watch",
    "start:types": "tsc --emitDeclarationOnly --declaration --declarationDir dist/types --watch",
    "prebuild": "npm run clean",
    "build": "npm run build:types && npm run build:main",
    "build:main": "cross-env NODE_ENV=production build-next-libs",
    "build:types": "tsc --emitDeclarationOnly --declaration --declarationDir dist/types"
  },
  "sideEffects": false,
  "dependencies": {
    "@next-core/http": "^1.0.0"
  },
  "devDependencies": {
    "@next-core/build-next-libs": "^1.0.0"
  }
}
",
  ],
]
`;

exports[`loadTemplate should work for public scope 1`] = `
[
  [
    "/tmp/sdk/src/wrapper.ts",
    "export interface ResponseBodyWrapper<T> {
  code: number;
  data: T;
  error?: string;
  message?: string;
}

export interface ResponseListWrapper<T> {
  list: T[];
  page: number;
  page_size: number;
  total: number;
}
",
  ],
  [
    "/tmp/sdk/tsconfig.json",
    "{
  "extends": "../../tsconfig.module.json",
  "include": ["./src"]
}
",
  ],
  [
    "/tmp/sdk/package.json",
    "{
  "name": "@next-api-sdk/cd-sdk",
  "version": "1.2.3",
  "description": "API SDK for Cd",
  "homepage": "https://github.com/easyops-cn/next-core/tree/master/sdk/cd-sdk",
  "license": "GPL-3.0",
  "type": "module",
  "main": "dist/cjs/index.js",
  "module": "dist/esm/index.js",
  "typings": "dist/types/index.d.ts",
  "files": [
    "dist",
    "contracts.json"
  ],
  "exports": {
    ".": {
      "types": "./dist/types/index.d.ts",
      "import": "./dist/esm/index.js",
      "require": "./dist/cjs/index.js"
    },
    "./package.json": "./package.json"
  },
  "scripts": {
    "clean": "rimraf dist",
    "prestart": "npm run clean",
    "start": "concurrently -k -n types,main \\"npm run start:types\\" \\"npm run start:main\\"",
    "start:main": "cross-env NODE_ENV=development build-next-libs --watch",
    "start:types": "tsc --emitDeclarationOnly --declaration --declarationDir dist/types --watch",
    "prebuild": "npm run clean",
    "build": "npm run build:types && npm run build:main",
    "build:main": "cross-env NODE_ENV=production build-next-libs",
    "build:types": "tsc --emitDeclarationOnly --declaration --declarationDir dist/types"
  },
  "sideEffects": false,
  "dependencies": {
    "@next-core/http": "^1.0.0"
  },
  "devDependencies": {
    "@next-core/build-next-libs": "^1.0.0"
  }
}
",
  ],
]
`;
